File: Dynamic-Programming/518. Coin Change 2/coinChange2.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        n = amount + 1
        dp = [0 for _ in range(n)]
        dp[0] = 1
        
        for coin in coins:
            for currency in range(n):
                if currency - coin >= 0:
                    dp[currency] = dp[currency] + dp[currency - coin]
        
        return dp[-1]            

class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        dp = [0 for _ in range(amount + 1)]
        dp[0] = 1
        
        for coin in coins:
            for i in range(amount + 1):
                if i - coin >= 0:
                    dp[i] = dp[i - coin] + dp[i]
        
        return dp[-1]

File: Dynamic-Programming/518. Coin Change 2/coinChange2.js
var change = function(amount, coins) {
    let arr = Array(amount + 1).fill(0);
    arr[0] = 1;
    
    for (const coin of coins) {
        for (let i = 1; i < arr.length; i++) {
            if (i - coin >= 0) {    
                // arr[i - coin] choose the coin
                // arr[i] not choose the coin, stay as current amount
                arr[i] = arr[i - coin] + arr[i];
            } 
        }
    }
    
    // console.log(arr);
    return arr[arr.length - 1];
};

/**
 * @param {number} amount
 * @param {number[]} coins
 * @return {number}
 */
 var change = function(amount, coins) {
    let dp = new Array(amount + 1).fill(0)
    dp[0] = 1
    
    for (const coin of coins) {
        for (let i = 1; i <= amount; i++) {
            if (i - coin >= 0) {
                dp[i] = dp[i - coin] + dp[i]
            }
        }
    }
    
    return dp[amount]
};

